code/modules/sound/sound_emitter.dm

```diff
@@ -0,0 +1,241 @@
+/datum/sound_emitter
+	var/atom/source
+	var/list/sounds = list()
+	var/active_key = null
+	var/channel = null
+	var/list/hearers = list()
+	var/range = 7
+
+	var/debug = FALSE
+
+/datum/sound_emitter/New(atom/s)
+	..()
+	source = s
+	range = world.view
+	SSsounds.register(src)
+
+/datum/sound_emitter/Destroy()
+	if (sounds)
+		sounds.Cut()
+		sounds = null
+	if (channel && channel != null)
+		stop()
+		release_channel()
+	hearers.Cut()
+	source = null
+	active_key = null
+	SSsounds.unregister(src)
+	. = ..()
+
+/datum/sound_emitter/proc/add(sound/s, var/key)
+	if (!s || !istype(s, /sound))
+		return
+	if (key && (key in sounds))
+		return
+	s.atom = source
+	s.environment = -1 // byond bug(?) if you set this to anything else, it will permanently set the channel environment to it
+	s.transform = matrix(1, 0, 0, 0, 1, 0) //dont think theres a good reason for this to be anything else
+	sounds[key] = s
+
+/datum/sound_emitter/proc/play(key)
+	var/sound/S = sounds[key]
+	if (!S)
+		world.log << "Sound emitter play called for key [key] on channel [channel], but sound does not exist."
+		return
+
+	if (S.repeat == 1)
+		// looping sounds need channel reservation so they can be stopped later
+		if (!channel)
+			channel = sound_channel_manager.reserve_channel(src)
+			if (!channel)
+				world.log << "Sound emitter could not reserve channel for sound [S.file]."
+				return
+		SSsounds.register_repeating_emitter(src)
+		S.channel = channel
+		active_key = key
+		// dispatch to hearers handled by SSsounds
+	else
+		// non-looping sounds do not need a channel reservation
+		// just send it
+		send_nearby_norepeat(S)
+
+/datum/sound_emitter/proc/apply_env_effects(sound/s)
+
+	// if an emitter is behind glass in a spaced area, listener shouldnt magically hear it in non-spaced nearby area
+	var/p_effect = turf_volume_coeff(s.atom)
+	s.volume *= p_effect
+
+	return s
+
+/datum/sound_emitter/proc/apply_player_effects(sound/s, var/mob/player)
+
+	if (player.ear_deaf > 0)
+		s.volume = s.volume / (1 + player.ear_deaf)
+
+	// loosely simulate some obstruction muffling the sound
+	if (!(source in view(range, player))) // TODO cache this
```

Comment:
As I mentioned above; adding a wait to each sound could make this less costly in exchange for accuracy. Using a cache also works but will also lose accuracy. Either works.